### ISSUE TITLE
Security Implications of JS Functions

### ISSUE THREAD

**Issue by cycomachead** [jmoenig/Snap#1754]
We should have a talk about how to handle this as be we can. 

Before we talk about solutions, I think we should list what types of activity we are most concerned about.

Me:
The problem happens when someone shares are project to the cloud and there is a JSFunction that does something unexpected that another user executes. I don't have as much of an issue with people shooting themselves in the foot by causing Snap! to hang or crash, though it'd be nice to make this as seamless as possible.

**Comment by jmoenig** [comment]
The issue is sniffing user credentials and modifying or deleting projects in the cloud account of a user that simply opens / runs such a project. That's the biggie.

**Comment by jmoenig** [comment]
Another issue is altering the environment by loading libraries, such as the Leapmotion library or - worse - the bignum library, that won't go away if you then open another project (only after you reload Snap, which is why we can't allow scripts that always run whenever you open Snap).

**Comment by brianharvey** [comment]
Clearly if you load someone else's project, the first time it calls JSFunction a warning dialog should pop up.  If you're really worried about it you can put in one of those really annoying delays before the OK button appears.
But, you know, if you copy someone's startup project without examining it, you deserve to lose.
PS Ha ha, I was just thinking this morning that I could load bignums in my startup file!  :-)

**Comment by brianharvey** [comment]
P.S.  I'm pretty sure I read somewhere recently something about how to handle login securely without keeping unencrypted passwords in local memory.  But I guess at a school, a bad guy could leave a modified Snap<i>!</i> running on a shared computer.  That might not be fixable, since the bad guy could patch out the hypothetical fix.

**Comment by cycomachead** [comment]
> P.S. I'm pretty sure I read somewhere recently something about how to handle login securely without keeping unencrypted passwords in local memory.

Not sure what you mean by "Local memory" but there's a well known practice, secure HTTPOnly cookies that are sent by a server and not readable via JS. 

> But, you know, if you copy someone's startup project without examining it, you deserve to lose.

I don't think so. Not only is the argument applied to other behaviors extremely problematic, but given that we start shared projects in an full-screen playable UI, which shows no code, we are not setting up the expectation that people need to review code. IMO, this goes contrary to everything about Snap! being a safe place for people to explore.

So the issues:
- Accessing user credentials: This is possible to prevent. (Secure HTTPOnly cookies)
- Reseting the Snap! Environment on project load: This we can work around a few different ways.

Things I don't see good ways of 100% preventing:
- Accessing user info / acting on behalf of the user
- changing saved settings

Both these can be (mostly) solved by disabling access to JS APIs before we execute a JSFunction. I'm not quite sure exactly how to do it off hand, but you can un-define a function and then later re-define it after executing something. 

That still leaves room for a bit of a cat-and-mouse game.

possible solutions:
- as mentioned: a warning. 
- we don't serialize the Text of a JSFunction when you make a project public.

**Comment by jmoenig** [comment]
Okay, I was sloppy about "sniffing" user credentials. I wasn't actually talking about reading user credentials in a way that they can be reused by someone else, I was talking about the user *themselves* unwantingly and even unconscientiouly letting someone else's project they open modify or delete their projects stored in the cloud. This has been done.

**Comment by cycomachead** [comment]
Well, right now, with the way we track logins, it _is_ possible for someone to grab someone else's credentials. (They can't read the plaintext password, but they could re-use the hash from another machine.) So, this is a problem worth considering.

Anyway, I'm not quite sure I followed what you described. Are you saying that User A inadvertently gave User B access to A's account? If so I'd be curious to know how this happened through a JSFunction. I can think of a couple ways to do it, but they would (I think) be clear to the author that you're doing something somewhat dangerous.

**Comment by jmoenig** [comment]
No, I'm saying that if you're logged in and you run arbitrary JS - or let a project run arbitrary JS - that JS function can also send XMLHttpRequests to the cloud. It can only access your own account, but this it can. Hey, remember when @bl-nero showed us? We didn't even discuss this openly because it's so creepy.

**Comment by cycomachead** [comment]
Oh yeah - that makes sense. I was counting that as something I'm not 100% sure we can prevent - though I think it might be possible.

1. We can disable (temporarily) access to the `Cloud` module. That makes interacting with the cloud, difficult but not impossible.
2. In terms of the cloud: We could, in theory, make it such that no user could take a permanently destructive action against themselves by maintaining backups. That prevents a total loss of data, but would still allow a lot of stress.
3. Proxying XHRs.

I think we can write a proxy function that intercepts / wraps the native XHR. There are a couple things we could do:
* if we disable access to cookies / localStorage / wherever we store credentials, then the XHR function should be able to send the necessary data along with the request.
* We can proxy such that we check the URL that's passed to the function and kill any requests sent to a blacklist of URLs. 

I think it might actually be possible to prevent such things, but I'm not positive. Certainly, we can make it very difficult.

We could also blanket disallow XHRs and any other networking in JSFunction -- though I think that means we would need to re-build the web requests library with built-in Snap<em>!</em> blocks, since those are _very_ useful teaching tools and lead to exciting projects.

**Comment by brianharvey** [comment]
> This has been done.

Did I know about this?  I don't remember it.

I am doubtful about any strategy that involves fencing off bits of the code because that would set off an arms race that defenders can never win.  (We have to be perfect; bad guys just have to find one hole.)

This may also be an unwinnable strategy, but let me throw out this idea:  We maintain a library of blessed, digitally signed JSFunction blocks.  (Yes, this is taking on a big workload.)  We make it easy (and free) to submit a block for review.  You can write JSFunction blocks and use them yourself without review.  You can load someone else's project that contains blessed JSFunction blocks.  You can't load someone else's project that contains unblessed ones.

Of course this doesn't prevent "social engineering" attacks in which the bad guy persuades someone to run a JSFunction block sent by email or something -- it'd be painful to disallow importing individual blocks because that's how people make homebrew libraries.

**Comment by cycomachead** [comment]
Yeah. I listed reviewing code as a possible solution. IMO, it's the only way that actually prevents malicious activity. (You could also speed up this work by testing for various API usage and flagging code with different levels of suspicion.)

Nothing prevents bad actors here. Anyone can send someone a link to Snap! or host their own version. 

However, I'm not sure we need to be perfect. In the case of protecting user accounts, regular backups that a only cloud admin has access to would prevent a user from losing work/having their account destroyed. 

But in the more general case: it seems that aside from the local Snap settings (which we are completely in control of), the biggest danger comes from networking events and dynamically loading JS. 

We do want both networking and dynamic JS loading (like BigNums), so outright banning those features doesn't necessary make sense. However, it's possible to work around XHRs with native Snap! blocks but then we end up in the same situation where someone can just make a web request in an easier   way. 
 
The more I think about it:
I'm not sure I think think of a foolproof way to prevent users from JS, without completely banning the ability to insert HTML into the web page all together. 

There is a native web API called Web Workers that is somewhat of a sandbox within JavaScript. The challenge would be getting a script executed in a webworker to be able to talk back to Snap! and do thing s like control Sprites, which we definitely do want to support. 

-- 
Michael Ball
From my iPhone
michaelball.co

**Comment by ScratchOs** [comment]
you could use something like [this](https://github.com/asvd/jailed/) which will allow you to only allow a certain api to jsfunctions

**Comment by bl-nero** [comment]
Just FYI, I once prepared a comprehensive proposal of fixing this problem:
https://docs.google.com/document/d/1rDREAHSmB6KhbTF2rTWfytCNcqWEQ
gBPz9gHq47K3mE/edit?usp=sharing. Unfortunately, I ran out of free time to
implement it (sorry!), but there is still some PoC work done in
https://github.com/bl-nero/Snap--Build-Your-Own-Blocks (see branch
https://github.com/bl-nero/Snap--Build-Your-Own-Blocks/commits/security;
beware, it's a year-old code, I'm not sure in what state I left it). Feel
free to check it out/grab/use/inspire yourself with it.

**Comment by bl-nero** [comment]
@ScratchOs One problem with Jailed is that you first need to expose a strict set of APIs for your JS code; that would be very challenging in the Snap! case.

I think I agree that a review process (and a clear notification before opening a project that contains unverified code) would be probably a lowest-hanging fruit.

**Comment by cycomachead** [comment]
> I think I agree that a review process (and a clear notification before opening a project that contains unverified code) would be probably a lowest-hanging fruit.

While an alert is a low-hanging fruit, it can have adverse consequences. A warning that users can't be expected to fairly judge its significance could be as bad as not having one at all.

As far as a review process - that requires a new cloud API, which seems a bit out of reach for a while.

**Comment by brianharvey** [comment]
On 5/31/17 1:56 AM, Michael Ball wrote:
> While an alert is a low-hanging fruit, it can have adverse consequences. A 
> warning that users can't be expected to fairly judge its significance could 
> be as bad as not having one at all.
Right, that's why I would propose that Snap! just plain refuse to load someone 
else's shared project with a non-blessed JS Function block.
> As far as a review process - that requires a new cloud API, which seems a bit out of reach for a while.
>
Why?  We don't have to store the blessed hashes at Miosoft; we can store them 
at Berkeley.  The rejection of projects happens in the front end at load time.

**Comment by cycomachead** [comment]
Well I suppose we can build our own thing, but it seems to me like this is info best stored with the project file. That way when you load a project it's already known whether not it's blessed and you don't slow things down. 

Reviewing every single JSFunction isn't quite enough. For example, our Web APIs in the libraries are safe functions but they can be used at runtime to be dangerous. (it's possible for someone to use the SnapCloud API using the web requests library). 
IMO, the review needs to be at the project level. Any time a project is made public, or a public project is updated, then it can be added to the reviewals queue. Projects that are likely safe or likely dangerous could be flagged appropriately. 

-- 
Michael Ball
From my iPhone
michaelball.co

**Comment by bl-nero** [comment]
That's a very good point, Michael.

**Comment by merlin2v** [comment]
Hello I am no expert but what about blocking connections to the snap database except through a specific JSBlock API that makes only one way to access the database and when it is used then a message will prompt up asking for permission to delete add create projects. however all other ways of entering the snap cloud is blocked when trying to access it from inside a JSBlock. 

example '
`promptCloud("GET","PARAMS")` → before it can get what is specified by the 'params' it has to prompt for permission

**Comment by cycomachead** [comment]
We could probably wall-off the functions used in Snap!'s cloud API library, but we _don't_ want to prevent people want using/accessing the native XMLHttpRequest function. If someone can access this, then they can essentially do whatever they want as a user. 

Maybe, it's enough to block off accessing credentials (i.e. Local Storage / cookies) from the JS Function block. However, this is still pretty tricky to do since everything in JS is global.
With a clever back-end API you might be able to watermark requests coming from a JS Function and restrict that to only certain properties, but you'd have to make sure that someone can't write code to remove the watermark. 

Plus, at a larger level, we don't want things like botnets, or people trying to just be disruptive in creative ways I can't think of right now.

**Comment by joefarebrother** [comment]
The issue with reviewing individual JSfunction blocks is that Js is very hard to audit, so a hole could easily make it past review. 

For example, suppose someone made a dictionary library that used JS objects internally. There would be 3 blocks, representing the following functions:

````JS
function make() {return {}}
function get(dict, key) {return dict[key]}
function set(dict, key, val) {dict[key] = val}
````

Those look innocent enough so might pass review. 
But then someone could import this library and use it to get `({})['constructor']['constructor']` from within Snap!, which is the `Function` constructor. Which can then be CALLed to build arbitrary, unreviewed functions, bypassing the whole process.
